lib/steps/images.js

```diff
@@ -0,0 +1,278 @@
+ *
+ * Input args:
+ * - `args.svcName`: The SAPI service name. This determines the appropriate
+ *   image name.
+ * - `args.channelArg`: Optional. The updates.joyent.com channel to use if
+ *   querying updates.joyent.com. It defaults to the default configured
+ *   channel (per `sdcadm channel`).
+ * - `args.imgArg`: must be one of:
+ *      - "latest": to find the latest available image on updates.joyent.com
+ *      - "current": the find the latest image already installed in the DC's
+ *        local IMGAPI, if any
+ *      - a UUID: to find that particular image
+ *      - a version string: to find that particular image version
+ * - `args.sdcadm`: The SdcAdm object for this run.
+ *
+ * Output args:
+ * - `args.svcImg`: The found image manifest.
+ * - `args.needToDownloadImg`: Set to true if the image is not already in the DC
+ *   and needs to be downloaded from updates.joyent.com.
+ * - `args.channel`: Maybe. If updates.joyent.com was queried (i.e. the image
+ *   was not found locally), then this will be set to `args.channelArg` or
+ *   resolved to the default channel.
+ *
+ * If no matching image can be found, this callback with an error.
+ */
+function findSvcImg(args, cb) {
+    assert.string(args.svcName, 'args.svcName');
+    assert.optionalString(args.channelArg, 'args.channelArg');
+    assert.string(args.imgArg, 'args.imgArg');
+    assert.object(args.sdcadm, 'args.sdcadm');
+
+    const sdcadm = args.sdcadm;
+    const context = {};
+
+    const imgName = sdcadm.config.imgNameFromSvcName[args.svcName];
+    if (!imgName) {
+        cb(new VError('could not determine image name for service "%s"',
+            args.svcName));
+        return;
+    }
+
+    vasync.pipeline({
+        arg: context,
+        funcs: [
+            // Note: It should be possible to avoid querying for the channel
+            // in some cases (e.g. if imgArg=current), but for now we just get
+            // it everytime.
+            function getChannel(ctx, next) {
+                if (!args.channelArg) {
+                    sdcadm.getDefaultChannel(function (err, channel) {
+                        if (err) {
+                            next(err);
+                            return;
+                        }
+                        ctx.channel = channel;
+                        next();
+                    });
+                } else {
+                    ctx.channel = args.channelArg;
+                    next();
+                }
+            },
+
+            // imgArg=='current' means look in the local IMGAPI.
+            function findImg_Current(ctx, next) {
+                if (args.imgArg !== 'current') {
+                    next();
+                    return;
+                }
+
+                sdcadm.imgapi.listImages({
+                    name: imgName
+                }, function (err, imgs) {
+                    if (err) {
+                        next(err);
+                    } else if (imgs && imgs.length > 0) {
```

Comment:
Wouldn't be fine with `images.length` on the conditional?.

Also, I personally prefer to return as soon as possible and avoid multiple conditional branches. For example, on this case:

```
if (err) {
    next(err)
    return;
}
if (!imgs || !imgs.length) {
    next(new errors.UpdateError(format(
        'no "%s" image found in this DC\'s IMGAPI',
        imgName)));
    return;
}
// Assumption: these are already sorted by published_at.
ctx.svcImg = imgs[imgs.length - 1];
ctx.needToDownloadImg = false;
next();

```

As commented, this is just a personal preference but, for cases where the intermediate branch becomes more complex, I find it more clear this way. The next function `finding_Latest` is also an example of the same thing. Note you do the `if(imgs  && imgs.length)` check there.

Reply:
> Wouldn't be fine with images.length on the conditional?.

Yes. Either is fine.  Some languages (I think Rust is an example) require that you do a numeric comparator for this.  So I am letting some Rust-learning show through.  I can change if you feel strongly?

> Also, I personally prefer to return as soon as possible and avoid multiple conditional branches. ...


I agree if the non-error case is at all long. For better or worse, I tend to balance that with a desire to have one and only one "return" in a function: at the end.  That leads to my style sometimes being:

```
 if (any error case) {
     next(err);
    return;
}
// ... handle the non-error case
```

and sometimes:

```
if (error case 1) {
      // small error handling, call next(...)
} else if (error case 2) {
     // small error handling, call next(...)
} else {
    // *small* handling the non-error case, call next(...)
}
```


I can change if you feel strongly. Note that I was largely translating what was already in lib/procedures/add-service.js to refactor this, and I believe it was using the latter style.